### PR TITLE
feat: add authentication and authorization using keycloak

### PR DESCRIPTION
**Pull request by HarikrishnanBalagopal**
Signed-off-by: Harikrishnan Balagopal <harikrishmenon@gmail.com>

### CODE REVIEW

File: .dockerignore [hunk 1]
```diff
@@ -15,7 +15,10 @@
...
 samples
 test_data
-bin
+bin/
 _dist
 coverage.txt
 workspace/
+my-sessions/
+# dist/
+# assets/swagger/
```
Comment (ashokponkumar) [inline .dockerignore:24]: Would be good to enable them or delete them

File: Dockerfile [hunk 2]
```diff
@@ -45,10 +45,20 @@ RUN make build
 
 # Run image
 FROM quay.io/konveyor/move2kube:${VERSION}
+WORKDIR /move2kube-api
 ENV MOVE2KUBE_PLATFORM="${MOVE2KUBE_PLATFORM}:api-dockerfile"
+
 # Install move2kube-api
 COPY --from=build_base /go/bin/move2kube-api /bin/move2kube-api
+# Copy static files for Move2Kube UI
+COPY ./dist ./dist
```
Comment (ashokponkumar) [inline Dockerfile:54]: I assume this will move to the move2kube-ui or move2kube-aio image

File: Dockerfile [hunk 3]
```diff
@@ -45,10 +45,20 @@ RUN make build
...
+
+# for openshift
+RUN chmod -R 775 /move2kube-api
+RUN chown -R 12345:root /move2kube-api
+USER 12345
```
Comment (ashokponkumar) [inline Dockerfile:59]: Instead we can change it to be executable by any user

File: cmd/move2kubeapi/move2kubeapi.go [hunk 4]
```diff
@@ -17,45 +17,111 @@ limitations under the License.
...
+	rootCmd.Flags().String("m2k_server_client_id", "m2k-server", "The OAuth 2.0 client id for the server side.")
+	rootCmd.Flags().String("m2k_server_client_secret", "8a1340ff-de5d-42a0-8b40-b6239c7cfc58", "The OAuth 2.0 client secret for the server side.")
+	rootCmd.Flags().String("default_resource_id", "b4d9b0fd-ffdb-4533-9536-5c315af07352", "Resource id on the Keycloak server.")
+	// CloudEvents
+	rootCmd.Flags().Bool("cloud_events_enabled", false, "Enable CloudEvents reporting.")
+	rootCmd.Flags().String("cloud_events_endpoint", "", "Endpoint where CloudEvents are reported.")
+	rootCmd.Flags().String("cloud_events_access_token", "", "Access token to use when reporting CloudEvents.")
+	rootCmd.Flags().String("cloud_events_spec_version", "1.0", "Version of the CloudEvents spec.")
+	rootCmd.Flags().String("cloud_events_type", "", "Type of the CloudEvents event.")
+	rootCmd.Flags().String("cloud_events_subject", "move2kube-api", "Subject to use when reporting CloudEvents.")
```
Comment (ashokponkumar) [inline cmd/move2kubeapi/move2kubeapi.go:76]: Seems like hyphen is the recommended practise (https://www.gnu.org/software/libc/manual/html_node/Argument-Syntax.html). We have to check how viper reacts to it though (https://stackoverflow.com/questions/61920435/map-flags-with-words-separated-by-dashes-to-camel-case-config)

File: Dockerfile [hunk 5]
```diff
@@ -45,10 +45,20 @@ RUN make build
...
+# for openshift
+
 # Start app
-WORKDIR /workspace
 EXPOSE 8080
-CMD move2kube-api -p 8080
+CMD ["move2kube-api", "--log_level", "info", "--static_files_dir", "dist"]
```
Comment (ashokponkumar) [inline Dockerfile:64]: We need to pass the port number, else it will end up with a random port number

File: Dockerfile [hunk 6]
```diff
@@ -45,10 +45,20 @@ RUN make build
 
 # Run image
 FROM quay.io/konveyor/move2kube:${VERSION}
+WORKDIR /move2kube-api
```
Comment (ashokponkumar) [inline Dockerfile:48]: why not use `workspace` or `workspaces`. The image by itself is for move2kube use only anyway.

File: assets/assets.go [hunk 7]
```diff
@@ -14,11 +14,11 @@ See the License for the specific language governing permissions and
...
-type AppMetadata struct {
-	Node  string `yaml:"node"`
-	URL   string `yaml:"url"`
-	Debug bool   `yaml:"debug"`
-}
+import (
+	"embed"
+)
+
+//go:embed swagger/*
```
Comment (ashokponkumar) [inline assets/assets.go:23]: Lets put the openapi yaml file alone using embed. For swagger files, we can possibly embed using the react (https://www.npmjs.com/package/swagger-ui-react) in our UI code.

File: cmd/move2kubeapi/move2kubeapi.go [hunk 8]
```diff
@@ -17,45 +17,111 @@ limitations under the License.
 package main
 
 import (
-	log "github.com/sirupsen/logrus"
-	"github.com/spf13/cobra"
-	"github.com/spf13/viper"
+	"log"
```
Comment (ashokponkumar) [inline cmd/move2kubeapi/move2kubeapi.go:20]: It should be logrus

File: internal/move2kubeapi/handlers/workspaces.go [hunk 9]
```diff
@@ -0,0 +1,261 @@
...
+	if err := json.NewEncoder(w).Encode(workspaces); err != nil {
+		logrus.Errorf("failed to send the workspaces json in the response. Error: %q", err)
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+	logrus.Debug("HandleListWorkspaces end")
+}
+
+func HandleCreateWorkspace(w http.ResponseWriter, r *http.Request) {
+	logrus.Debug("HandleCreateWorkspace start")
```
Comment (ashokponkumar) [inline internal/move2kubeapi/handlers/workspaces.go:94]: Do we want to make all these messages as trace?

File: internal/move2kubeapi/move2kubeapi.go [hunk 10]
```diff
@@ -17,430 +17,199 @@ limitations under the License.
...
-	name := mux.Vars(r)[nameRouteVar]
-	plan := r.FormValue("plan")
 
-	artifactName := name + "-" + cast.ToString(time.Now().Unix())
-	log.Infof("Artifact Name:%s", artifactName)
+	// static routes
+	// swagger UI
+	swaggerDir, _ := fs.Sub(assets.SwaggerUI, "swagger")
+	router.Handle("/swagger", http.RedirectHandler("/swagger/", http.StatusMovedPermanently))
+	router.PathPrefix("/swagger/").Handler(http.StripPrefix("/swagger", http.FileServer(http.FS(swaggerDir)))).Methods("GET")
```
Comment (ashokponkumar) [inline internal/move2kubeapi/move2kubeapi.go:141]: We can consider serving this as part of the UI

**Review by ashokponkumar — CHANGES_REQUESTED** [review]
Nice work Hari! Just a few suggestions...

File: .dockerignore (hunk 1 above)
Comment (HarikrishnanBalagopal) [reply .dockerignore:24]: done

File: Dockerfile (hunk 2 above)
Comment (HarikrishnanBalagopal) [reply Dockerfile:54]: done

File: Dockerfile (hunk 3 above)
Comment (HarikrishnanBalagopal) [reply Dockerfile:59]: done

File: Dockerfile (hunk 5 above)
Comment (HarikrishnanBalagopal) [reply Dockerfile:64]: done

File: Dockerfile (hunk 6 above)
Comment (HarikrishnanBalagopal) [reply Dockerfile:48]: done

File: assets/assets.go (hunk 7 above)
Comment (HarikrishnanBalagopal) [reply assets/assets.go:23]: done

File: internal/move2kubeapi/move2kubeapi.go (hunk 10 above)
Comment (HarikrishnanBalagopal) [reply internal/move2kubeapi/move2kubeapi.go:141]: done

File: cmd/move2kubeapi/move2kubeapi.go (hunk 4 above)
Comment (HarikrishnanBalagopal) [reply cmd/move2kubeapi/move2kubeapi.go]: done

File: cmd/move2kubeapi/move2kubeapi.go (hunk 8 above)
Comment (HarikrishnanBalagopal) [reply cmd/move2kubeapi/move2kubeapi.go:20]: done

File: internal/move2kubeapi/handlers/workspaces.go (hunk 9 above)
Comment (HarikrishnanBalagopal) [reply internal/move2kubeapi/handlers/workspaces.go:94]: done

**Review by ashokponkumar — APPROVED** [review]

**Comment by ashokponkumar** [comment]
Seems like the build is failing. Once the build passes we can merge.

File: Dockerfile [hunk 11]
```diff
@@ -45,10 +45,16 @@ RUN make build
...
+# for openshift
+RUN chown -R :root /move2kube-api
+RUN chmod -R 770 /move2kube-api
+# for openshift
+
 # Start app
-WORKDIR /workspace
 EXPOSE 8080
-CMD move2kube-api -p 8080
+CMD ["move2kube-api", "--port", "8080", "--log_level", "info"]
```
Comment (ashokponkumar) [inline Dockerfile:60]: Isn't it supposed to be hyphen?

**Review by ashokponkumar — CHANGES_REQUESTED** [review]

File: internal/move2kubeapi/handlers/roles.go [hunk 12]
```diff
@@ -0,0 +1,239 @@
...
+func HandleReadRole(w http.ResponseWriter, r *http.Request) {
+	logrus.Trace("HandleReadRole start")
+	accessToken, err := common.GetAccesTokenFromAuthzHeader(r)
+	if err != nil {
+		logrus.Debugf("failed to get the access token from the request. Error: %q", err)
+		w.Header().Set(common.AUTHENTICATE_HEADER, common.AUTHENTICATE_HEADER_MSG)
+		w.WriteHeader(http.StatusUnauthorized)
+		return
+	}
+	roleId := mux.Vars(r)["role-id"]
```
Comment (ashokponkumar) [inline internal/move2kubeapi/handlers/roles.go:132]: We seem to be using role-id in more than one place, better to use a const to make sure they don't end up diverging.

**Review by ashokponkumar — COMMENTED** [review]
Few nit. When you are fixing lint errors you can look at them

File: Dockerfile (hunk 11 above)
Comment (HarikrishnanBalagopal) [reply Dockerfile:60]: done

File: cmd/move2kubeapi/move2kubeapi.go [hunk 13]
```diff
@@ -17,45 +17,113 @@ limitations under the License.
...
+	configFilePath := viper.GetString("config")
+	logrus.Info("reading config from file at path", configFilePath)
+	viper.SetConfigFile(configFilePath)
+	if err := viper.ReadInConfig(); err != nil {
+		logrus.Fatalf("failed to read the config file at path %s . Error: %q", configFilePath, err)
+	}
+}
+
+func setupViper() {
+	viper.SetEnvPrefix("M2K")
```
[the remark this answers is not in the document]
Comment (HarikrishnanBalagopal) [reply cmd/move2kubeapi/move2kubeapi.go:99]: done

File: internal/move2kubeapi/handlers/handlers.go [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+
+	"github.com/konveyor/move2kube-api/internal/authserver"
+	"github.com/konveyor/move2kube-api/internal/common"
+	"github.com/konveyor/move2kube-api/internal/filesystem"
+	"github.com/konveyor/move2kube-api/internal/sessions"
+	"github.com/sirupsen/logrus"
+)
+
+const (
+	USER_SESSION_NAME = "m2k-ui-user"
```
[the remark this answers is not in the document]
Comment (HarikrishnanBalagopal) [reply internal/move2kubeapi/handlers/handlers.go]: done

File: internal/move2kubeapi/handlers/roles.go (hunk 12 above)
Comment (HarikrishnanBalagopal) [reply internal/move2kubeapi/handlers/roles.go]: done

File: internal/move2kubeapi/handlers/middleware.go [hunk 15]
```diff
@@ -0,0 +1,150 @@
...
+				w.Header().Set(common.AUTHENTICATE_HEADER, common.AUTHENTICATE_HEADER_MSG)
+				w.WriteHeader(http.StatusUnauthorized)
+				return
+			}
+			accessToken = session.Tokens.AccessToken
+			r.Header.Set(common.AUTHZ_HEADER, "Bearer "+accessToken)
+		}
+
+		// "GET" on "/workspaces" allows a user to list workspaces they have access to.
+		if (verb == "GET" && resPath == "/workspaces") ||
```
[the remark this answers is not in the document]
Comment (HarikrishnanBalagopal) [reply internal/move2kubeapi/handlers/middleware.go]: now it's only in one place

**Review by ashokponkumar — DISMISSED** [review]

**Comment by ashokponkumar** [comment]
We can use a comment like https://github.com/konveyor/move2kube/blob/main/assets/assets.go for embed variables too

**Review by ashokponkumar — APPROVED** [review]

**Pull request merged**